share/dkg/dkg.go

```diff
@@ -0,0 +1,660 @@
+// Dealer.
+//  NOTE: Doing that in vss.go would be possible but then the Dealer is always
+//  assumed to be a member of the participants. It's only the case here.
+type Deal struct {
+	// Index of the Dealer in the list of participants
+	Index uint32
+	// Deal issued for another participant.
+	Deal *vss.EncryptedDeal
+}
+
+// Response holds the Response from another participants as well as the index of
+// the targeted Dealer.
+type Response struct {
+	// Index of the Dealer for which this response is for
+	Index uint32
+	// Response issued from another participant.
+	Response *vss.Response
+}
+
+// Justification holds the Justification from a Dealer as well as the index of
+// the Dealer in question.
+type Justification struct {
+	// Index of the Dealer who answered with this Justification
+	Index uint32
+	// Justification issued from the Dealer.
+	Justification *vss.Justification
+}
+
+// SecretCommits is sent during the distributed public key reconstruction phase,
+// basically a Feldman VSS scheme.
+type SecretCommits struct {
+	// Index of the Dealer in the list of participants
+	Index uint32
+	// Commitments generated by the Dealer
+	Commitments []abstract.Point
+	// SessionID generated by the Dealer tied to the Deal.
+	SessionID []byte
+	// Signature from the Dealer
+	Signature []byte
+}
+
+// ComplaintCommits is sent if the secret commitments revealed by a peer are not
+// valid.
+type ComplaintCommits struct {
+	// Index of the Verifier _issuing_ the ComplaintCommit
+	Index uint32
+	// DealerIndex being the index of the Dealer who issued the SecretCommits
+	DealerIndex uint32
+	// Deal that has been given from the Dealer (at DealerIndex) to this node
+	// (at Index)
+	Deal *vss.Deal
+	// Signature made by the verifier
+	Signature []byte
+}
+
+// ReconstructCommits holds the information given by a participants who reveals
+// the deal received from a peer that have received a ComplaintCommits.
+type ReconstructCommits struct {
+	// Id of the session
+	SessionID []byte
+	// Index of the verifier who received the deal
+	Index uint32
+	// DealerIndex is the index of the dealer who issued the Deal
+	DealerIndex uint32
+	// Share contained in the Deal
+	Share *share.PriShare
+	// Signature over all over fields generated by the issuing verifier.
+	Signature []byte
+}
+
+// DistKeyGenerator is the struct that runs the DKG protocol.
+type DistKeyGenerator struct {
+	suite abstract.Suite
+
+	index uint32
+	long  abstract.Scalar
+	pub   abstract.Point
+
+	participants []abstract.Point
+
+	t int
+
+	dealer    *vss.Dealer
+	verifiers map[uint32]*vss.Verifier
+
+	// list of commitments to each secret polynomials
+	commitments map[uint32]*share.PubPoly
+
+	// map of deals collected to reconstruct the full polynomial of a dealer
+	// key is index of the dealer. Once there is enough ReconstructCommits
+	// struct, the dkg will re-construct the polynomial and stores it into
+	// the list of commitments.
```

Comment:
> s/stores/store/
It's *this* dkg that will reconstruct the secret, not every dkg since we don't know if they got the share or not. I'll write "this dkg" instead of "the dkg".